gh-119349: Add ctypes.util function to list loaded shared libraries (#122946)
python/cpython: Doc/library/ctypes.rst

module instead of using :func:`~ctypes.util.find_library` to locate the library at runtime.

.. _ctypes-listing-loaded-shared-libraries:

Listing loaded shared libraries
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

When writing code that relies on code loaded from shared libraries, it can be
useful to know which shared libraries have already been loaded into the current
process.

The :mod:`!ctypes.util` module provides the :func:`~ctypes.util.dllist` function,
which calls the different APIs provided by the various platforms to help determine

> which shared libraries have already been loaded into the current process.

Reviewer: Tiny little nitpick: we could add an example here.

Author: I added an example and a slightly longer description to give it context

module instead of using :func:`~ctypes.util.find_library` to locate the library at runtime.

.. _ctypes-listing-loaded-shared-libraries:

Listing loaded shared libraries
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

When writing code that relies on code loaded from shared libraries, it can be
useful to know which shared libraries have already been loaded into the current
process.

The :mod:`!ctypes.util` module provides the :func:`~ctypes.util.dllist` function,
which calls the different APIs provided by the various platforms to help determine
which shared libraries have already been loaded into the current process.

> 
> .. data:: dllist()
>    :module: ctypes.util
>    :noindex:
> 
>    Try to provide a list of paths of the shared libraries loaded into the current
>    process.  These paths are not normalized or processed in any way. If the
>    function is not able to determine the list of loaded libraries (either because
>    the current platform does not support it, or because an error occurred), it
>    returns ``None``.
> 
> The exact functionality is system dependent.

Reviewer 2: Let's only have this in one place; users can follow the link.
```suggestion
```

   function in the same library that allocated the memory.

.. function:: dllist()
   :module: ctypes.util

   Try to provide a list of paths of the shared libraries loaded into the current
   process.  These paths are not normalized or processed in any way. If the
   function is not able to determine the list of loaded libraries (either because
   the current platform does not support it, or because an error occurred), it
   returns ``None``.

   The exact functionality is system dependent.

> 

Reviewer: ```suggestion
   .. versionadded:: 3.14
```

   function in the same library that allocated the memory.

.. function:: dllist()
   :module: ctypes.util

   Try to provide a list of paths of the shared libraries loaded into the current
   process.  These paths are not normalized or processed in any way. If the
   function is not able to determine the list of loaded libraries (either because

>    the current platform does not support it, or because an error occurred), it

Reviewer: If the current platform doesn't support it, we should at least document *why*. Could you add an excerpt about the system needing libc to have `dl_iterate_phdr` in order for this to work?

Author: I could, though I could imagine such a thing going stale if specific support for another platform was added down the line, or some other API was used

Reviewer: Sure, and then we document that later on.

   .. availability:: Windows

.. function:: dllist()
   :module: ctypes.util

   Try to provide a list of paths of the shared libraries loaded into the current
   process.  These paths are not normalized or processed in any way.  The function
   can raise :exc:`OSError` if the underlying platform APIs fail.

>    The first element of the list is typically the path of the current executable
>    file.  The exact functionality is system dependent.

Reviewer 2: ```suggestion
   The exact functionality is system dependent.

   On most platforms, the first element of the list represents the current
   executable file. It may be an empty string.
```

   .. availability:: Windows

.. function:: dllist()
   :module: ctypes.util

   Try to provide a list of paths of the shared libraries loaded into the current
   process.  These paths are not normalized or processed in any way.  The function
   can raise :exc:`OSError` if the underlying platform APIs fail.
   The exact functionality is system dependent.

   On most platforms, the first element of the list represents the current
   executable file. It may be an empty string.

   .. availability:: Windows, macOS, iOS, glibc, BSD libc, musl

>    .. versionadded:: 3.14

Reviewer: We have this cool thing:
```suggestion
   .. versionadded:: next
```

   .. availability:: Windows

.. function:: dllist()
   :module: ctypes.util

   Try to provide a list of paths of the shared libraries loaded into the current
   process.  These paths are not normalized or processed in any way. If the
   function is not able to determine the list of loaded libraries (either because
   the current platform is not Windows, MacOS, or Linux and does not support the
   the ``dl_iterate_phdr`` API in libc, or because an error occurred), it
   returns ``None``.

   The exact functionality is system dependent.

> 

Reviewer 2: Please add `Availability` listing the platforms; ISTM should be
```
.. availability:: Windows, macOS, glibc, BSD libc
```
and perhaps `musl` too. The list is at `Doc/tools/extensions/availability.py`.

Author: I listed those you did. I _believe_ it would also work as-implemented on iOS, Android, and musl, but I have not verified these myself

Author: Testing in an alpine docker image (musl), the code here works _except_ that `find_library('c')` returns none there.
I can still get a handle on libc by using `ctypes.CDLL(None)` -- is this legit to do in the `util` code? 

I see it being used once in the ctypes documentation, but it seems absent from the existing cpython codebase.

Reviewer 2: For POSIX, it's fine. Updating the `ctypes.CDLL` docs is tracked in #78773.
